DarklyEventSourceTests/LDEventSourceTest.m

```diff
@@ -0,0 +1,221 @@
+    [eventSource onMessage:^(LDEvent *event) {
+        XCTAssertNotNil(event);
+        XCTAssertEqualObjects(event.event, @"put");
+        XCTAssertEqualObjects(event.data, putEventData);
+        XCTAssertEqual(event.readyState, kEventStateOpen);
+
+        [eventExpectation fulfill];
+    }];
+
+    for (NSString *eventStringPart in putEventStringParts) {
+        [eventSource URLSession:eventSource.session dataTask:eventSource.eventSourceTask didReceiveData:[eventStringPart dataUsingEncoding:NSUTF8StringEncoding]];
+    }
+
+    [self waitForExpectations:@[eventExpectation] timeout:2.0];
+    [OHHTTPStubs removeAllStubs];
+}
+
+-(void)testDidReceiveData_multipleCalls_randomParts {
+    NSString *dummyClientStreamHost = @"dummy.clientstream.launchdarkly.com";
+
+    NSString *putEventString = [NSString stringFromFileNamed:@"largePutEvent"];
+    NSString *putEventData = [[putEventString componentsSeparatedByString:@"data:"] lastObject];
+    putEventData = [putEventData substringToIndex:putEventData.length - 2]; //chop off the last 2 characters
+    NSArray *putEventStringParts = [putEventString splitIntoPartsApproximatelySized:1024];
+
+    [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest * _Nonnull request) {
+        return [request.URL.host isEqualToString:dummyClientStreamHost];
+    } withStubResponse:^OHHTTPStubsResponse * _Nonnull(NSURLRequest * _Nonnull request) {
+        return [OHHTTPStubsResponse responseWithData:[NSData data] statusCode:200 headers:nil];
+    }];
+
+    XCTestExpectation *eventExpectation = [self expectationWithDescription:@"LDEventSourceTest.testDidReceiveData_multipleCalls.eventExpectation"];
+
+    LDEventSource *eventSource = [LDEventSource eventSourceWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"https://%@", dummyClientStreamHost]] httpHeaders:nil];
+    [eventSource onMessage:^(LDEvent *event) {
+        XCTAssertNotNil(event);
+        XCTAssertEqualObjects(event.event, @"put");
+        XCTAssertEqualObjects(event.data, putEventData);
+        XCTAssertEqual(event.readyState, kEventStateOpen);
+
+        [eventExpectation fulfill];
+    }];
+
+    for (NSString *eventStringPart in putEventStringParts) {
+        [eventSource URLSession:eventSource.session dataTask:eventSource.eventSourceTask didReceiveData:[eventStringPart dataUsingEncoding:NSUTF8StringEncoding]];
+    }
+
+    [self waitForExpectations:@[eventExpectation] timeout:2.0];
+    [OHHTTPStubs removeAllStubs];
+}
+
+-(void)testDidReceiveData_extraNewLine {
+    NSString *dummyClientStreamHost = @"dummy.clientstream.launchdarkly.com";
+
+    NSString *putEventString = [NSString stringFromFileNamed:@"largePutEvent"];
+    NSMutableArray *putEventStringParts = [NSMutableArray arrayWithArray:[putEventString componentsSeparatedByString:@":\""]];
+    NSUInteger selectedIndex = arc4random_uniform((uint32_t)putEventStringParts.count - 1) + 1;
```

Comment:
I'm wary of using random in tests, as it makes the test non-deterministic, and if there is a specific case that fails, this will only catch it some of the time. Ideally, we'd just provide tests that represent the different possible cases.

Reply:
LD specifically asked for this kind of test.